### ISSUE TITLE
use of  pdb.set_trace() inside LightningModule

### ISSUE THREAD

**Issue by moabaom** [PyTorchLightning/pytorch-lightning#5984]
So being new to lightning, I have rather a basic question that I couldn't quite find the answer to in the docs.  So in regular pytorch I can use python's pdb module to pause the execution at any point and like check tensor dimensions.values, etc : 

from pdb import set_trace
set_trace()

However I seem not be able to do so with lightning, whether in forward() or training_step() or any other method in the lightning module I can pause the execution at any point (obviously without having to go into lightning's internal modules). Is this supposed to be not possible or what am i missing?

**Comment by akihironitta** [comment]
@Moadab-AI Could I ask in what case you want to pause executions outside LightningModule? If you put `from pdb import set_trace; set_trace()` in LightningModule's methods, `training_step()` for example, you can still check tensor sizes/values when building models, right?

**Comment by adam-mehdi** [comment]
I have the same issue; I am debugging a complicated GAN and want to halt execution inside of `LightningModule.forward()` . Is there a way to do this?

**Comment by akihironitta** [comment]
Yes, you can stop execution by placing `set_trace()` in `LightningModule.forward()` just like when you debug normal Python scripts.

```python
class YourGAN(LightningModule):
    ...
    def forward(self, batch):
        ...
        from pdb import set_trace; set_trace()
        ...
    ...
```

**Comment by adam-mehdi** [comment]
Thanks for the response!

**Comment by akihironitta** [comment]
@adam-mehdi Note that `forward()` is optional and the method isn't executed unless you manually call it in `training_step` (or wherever).

Please also see https://pytorch-lightning.readthedocs.io/en/latest/starter/new-project.html.

> FORWARD vs TRAINING_STEP
>
> In Lightning we separate training from inference. The training_step defines the full training loop. We encourage users to use the forward to define inference actions. Of course, nothing is stopping you from using forward from within the training_step. It really comes down to your application. We do, however, recommend that you keep both intents separate.
> - Use forward for inference (predicting).
> - Use training_step for training.

**Comment by sounakdey** [comment]
@akihironitta if one wants to check the values in the validation step `imoprt pdb; pdb.set_trace()` though stops the execution but one cannot type commands to see what the value of the variable or check what would happen in a pythonic way.. is there a way to go about it

**Comment by sounakdey** [comment]
Ahh no I realized its the wandb logger that was creating the problem... yes I could see the console at the breakpoint it's just did not let me write things to debug

**Comment by maruoss** [comment]
I set the new breakpoint() in training_step and it stops after 2-3 c(ontinuations).
